### ISSUE TITLE
O(1, (n, oo)) is contained in O(sin(n), (n, oo))

### ISSUE THREAD

**Issue by gxyd** [sympy/sympy#10390]
```
>>> n = Symbol('n', integer=True, positive=True)
>>> e1 = O(1, (n, oo))
>>> e2 = O(sin(n), (n, oo))
>>> e2 in e1
True         # this is correct
>>> e1 in e2
True         # this should be false
```

**Comment by tokencolour** [comment]
Hello @gxyd,
I read [this](http://docs.sympy.org/dev/modules/series.html#order-terms).  I couldn't quite understand why O(x + x**2) is O(x). Is there a condition imposed on x?
```
>>> O(x+x**2)
O(x)
```
and what the second arg (the tuple (x, 0) or (x, oo)) means
```
>>> O(x + x**2, (x, 0))
O(x)

>>> O(x + x**2, (x, oo))
O(x**2, (x, oo))
```
Can you please recommend to me some reading material so that I can comprehend this?

**Comment by gxyd** [comment]
You can read this  https://en.wikipedia.org/wiki/Big_O_notation . I don't know the fix right now. 
In particular in mathematics, `BigO` is "not just" about taking limit behaviour at `oo`(which is often done in computer science) that is done specifying a tuple `(x, oo)`.  We can find the limiting behaviour at any general point as `x tends to a` using tuple `(x, a)`

**Comment by tokencolour** [comment]
Thank you!

**Comment by gxyd** [comment]
This will probably be solved after the implementation of `#TODO` left for  `AccumBounds.union` method. Since `limit(1/sin(n), (n, oo))` returns `<-oo, oo>` instead of `<-oo, -1> U <1, oo>`.

**Comment by tokencolour** [comment]
I debugged a little to find out that the reason why `limit(1/sin(n), (n, oo))` returns `<-oo, oo>` instead of `<-oo, -1> U <1, oo>` . It seemed to me that the problem is [here](https://github.com/sympy/sympy/blob/master/sympy/calculus/util.py#L561). I couldn't understand how this is related to `AccumBounds.union` being incompletely implemented. Can you please help me, @gxyd?

Also, I am interested to complete the implementation of `union` in `AccumulationBounds` . One test case I found, where the `union` fails is:
```
AccumBounds(-oo, -1).union(AccumBounds(0, oo))
```

**Comment by gxyd** [comment]
>  I couldn't understand how this is related to AccumBounds.union being incompletely implemented.

The actual answer (if you debugged it right) to the point you are mentioning is because of the same reason as [these piece of code](https://github.com/sympy/sympy/blob/master/sympy/calculus/util.py#L453-L457) (see the comments) on the same file.

> Also, I am interested to complete the implementation of union in AccumulationBounds

`@jksuom` also told me
> I was playing with the idea of switching the end points to indicate the complement.
Of course, that will make the arithmetic more complicated.
But unions would probably also lead to similar complications.

I don't have idea of a good solution to this.

**Comment by 0sidharth** [comment]
> implementation of `#TODO` left for `AccumBounds.union` method

Would introducing a new class `AccumulationBoundsUnion` be a good solution to this? This is how it is implemented for `Set` in `sets.py`, and we cannot represent something like `<-oo, -1> U <1, oo>` using the current `AccumulationBounds` class.

**Comment by anutosh491** [comment]
> > implementation of `#TODO` left for `AccumBounds.union` method
> 
> Would introducing a new class `AccumulationBoundsUnion` be a good solution to this? This is how it is implemented for `Set` in `sets.py`, and we cannot represent something like `<-oo, -1> U <1, oo>` using the current `AccumulationBounds` class.
 
yeah , exactly !! I was about to make a Pr for another issue , but eventually I realized this hasn't been done yet and the Pr wouldn't function as expected ! Not exactly sure about the best option here. @jksuom @oscarbenjamin Could you'll guide me through what could be a good approach for this issue ?

**Comment by jksuom** [comment]
Maybe `O(sin(n, (n, oo))` could be replaced by `O(1, (n, oo))`.

**Comment by anutosh491** [comment]
Will have to think about the pros and cons though . I see some use-cases but then the issue is that 
```
>>> O(sin(n) , (n, oo)) in O(1, (n, oo)) # Should give false though
True
 # After replacing 
>>> O(1, (n, oo)) in O(1, (n, oo))
True
```
Will this be helping ??

**Comment by anutosh491** [comment]
> Maybe `O(sin(n, (n, oo))` could be replaced by `O(1, (n, oo))`.

I feel we should still have a `union` class for `AccumBounds` like the one in sets maybe . Because it anyways affects all division operations involving `AccumBounds`
```
For eg
>>> 1/(AccumBounds(-1, 1))
AccumBounds(-oo, oo)
```
